src/Storages/MergeTree/KeyCondition.cpp

```diff
@@ -2274,25 +2295,24 @@ static BoolMask forAnyHyperrectangle(
             hyperrectangle[i] = Range::createWholeUniverseWithoutNull();
     }
 
-    BoolMask result = initial_mask;
-    result = result | callback(hyperrectangle);
+    auto result = BoolMask::combine(initial_mask, callback(hyperrectangle));
 
     /// There are several early-exit conditions (like the one below) hereinafter.
-    /// They are important; in particular, if initial_mask == BoolMask::consider_only_can_be_true
-    /// (which happens when this routine is called from KeyCondition::mayBeTrueXXX),
-    /// they provide significant speedup, which may be observed on merge_tree_huge_pk performance test.
-    if (result.isComplete())
+    /// They provide significant speedup, which may be observed on merge_tree_huge_pk performance test.
+    if (result.isComplete(initial_mask))
```

Comment:
This change I don't fully understand.
First of all, the logic has changed. Previously we used `result | callback(hyperrectangle)`, and now we use `BoolMask::combine(initial_mask, callback(hyperrectangle))`, which gives a different result.

Alos now `result.isComplete` takes `initial_mask` again as an argument. Is it needed here?

Reply:
Or, probably, there was a bug previously?
Than, let's make a separate fix and backport it.